File: solutions/lua/dnd-character/2/dnd-character.lua
local function modifier(input) return (input - 10) // 2 end
local function ability(scores)
  scores = scores or { math.random(1, 6), math.random(1, 6), math.random(1, 6), math.random(1, 6) }
  table.sort(scores, function(a, b) return a > b end)
  return scores[1] + scores[2] + scores[3]
end

local Character = {}
function Character:new(name)
  local constitution = ability()
  return {
    name = name,
    strength = ability(),
    dexterity = ability(),
    constitution = constitution,
    intelligence = ability(),
    wisdom = ability(),
    charisma = ability(),
    hitpoints = 10 + modifier(constitution),
  }
end

return { Character = Character, ability = ability, modifier = modifier }

File: solutions/clojure/dnd-character/2/src/dnd_character.clj
(ns dnd-character
  (:require [clojure.math :refer [floor-div]]))

(defn score-modifier [score]
  (floor-div (- score 10) 2))

(defn rand-ability []
  (->> (repeatedly 4 #(inc (rand-int 6)))
       sort
       rest
       (reduce +)))

(defn rand-character []
  (let [constitution (rand-ability)]
    (assoc (zipmap [:strength :dexterity :intelligence :wisdom :charisma]
                   (repeatedly 5 rand-ability))
           :constitution constitution
           :hitpoints (+ 10 (score-modifier constitution)))))
